llama_index/schema.py

```diff
@@ -119,3 +147,303 @@ def extra_info_str(self) -> Optional[str]:
             return None
 
         return "\n".join([f"{k}: {str(v)}" for k, v in self.extra_info.items()])
+
+
+@dataclass
+class BaseIndexObject(DataClassJsonMixin):
+    """Base Index Object.
+
+    Generic abstract interface for retrievable objects
+
+    """
+
+    _id: str = ""
+    content: Any = None
+    relationships: Dict[DataRelationship, Any] = field(default_factory=dict)
+    embedding: Optional[List[float]] = None
+    object_hash: Optional[str] = None
+    weight: float = 1.0
+
+    """"
+    metadata fields
+    - injected as part of the text shown to LLMs as context
+    - used by vector DBs for metadata filtering
+
+    This must be a flat dictionary, 
+    and only uses str keys, and (str, int, float) values.
+    """
+    metadata: Dict[str, Any] = field(default_factory=dict)
+
+    # attributes are additional fields that do not influence query/retrieval
+    attributes: Dict[str, Any] = field(default_factory=dict)
+
+    def __post_init__(self) -> None:
+        """Post init."""
+        # assign doc_id if not set
+        if not self._id:
+            self._id = get_new_id(set())
+        if self.object_hash is None:
+            self.object_hash = self._generate_object_hash()
+
+        if self.metadata is not None:
+            _validate_is_flat_dict(self.metadata)
+
+    @abstractmethod
+    def _generate_object_hash(self) -> str:
+        """Generate a hash to represent the object."""
+
+    @classmethod
+    @abstractmethod
+    def get_type(cls) -> str:
+        """Get Object type."""
+
+    @abstractmethod
+    def get_content(self) -> str:
+        """Get object content."""
+
+    def get_object_id(self) -> str:
+        """Get doc_id."""
+        if self._id is None:
+            raise ValueError("_id not set.")
+        return self._id
+
+    def get_object_hash(self) -> str:
+        """Get object_hash."""
+        if self.object_hash is None:
+            raise ValueError("object_hash is not set.")
+        return self.object_hash
+
+    @property
+    def object_id(self) -> str:
+        return self._id
+
+    @property
+    def is_id_none(self) -> bool:
+        """Check if doc_id is None."""
+        return self.is_id_none is None
+
+    @property
+    def is_content_none(self) -> bool:
+        """Check if text is None."""
+        return self.content is None
+
+    def get_embedding(self) -> List[float]:
+        """Get embedding.
+
+        Errors if embedding is None.
+
+        """
+        if self.embedding is None:
+            raise ValueError("embedding not set.")
+        return self.embedding
+
+    @property
+    def metadata_str(self) -> str:
+        """Extra info string."""
+        if self.metadata is None:
+            return ""
+
+        return "\n".join([f"{k}: {str(v)}" for k, v in self.metadata.items()])
```

Comment:
I do wonder if this is too opinionated. Is it possible for users to customize how they'd want to inject metadata into the text / how are we thinking about that?

Reply:
Hmm, metadata and content-level content templates maybe? lol